worldedit-core/src/main/java/com/fastasyncworldedit/core/history/changeset/AbstractChangeSet.java

```diff
@@ -372,25 +364,58 @@ public Future<?> addWriteTask(final Runnable writeTask, final boolean completeNo
                 } else {
                     int hash = t.getMessage().hashCode();
                     if (lastException.getAndSet(hash) != hash) {
-                        t.printStackTrace();
-                    }
-                }
-            } finally {
-                if (AbstractChangeSet.this.waitingCombined.decrementAndGet() <= 0) {
-                    synchronized (AbstractChangeSet.this.waitingAsync) {
-                        AbstractChangeSet.this.waitingAsync.notifyAll();
-                    }
-                    synchronized (AbstractChangeSet.this.waitingCombined) {
-                        AbstractChangeSet.this.waitingCombined.notifyAll();
+                        LOGGER.catching(t);
                     }
                 }
             }
         };
         if (completeNow) {
             wrappedTask.run();
-            return Futures.immediateCancelledFuture();
+            return Futures.immediateVoidFuture();
         } else {
-            return Fawe.instance().getQueueHandler().submit(wrappedTask);
+            CompletableFuture<?> task = new CompletableFuture<>();
+            queue.add(() -> {
+                wrappedTask.run();
+                task.complete(null);
+            });
+            // make sure changes are processed
+            triggerWorker();
+            return task;
+        }
+    }
+
+    private void triggerWorker() {
+        if (workerSemaphore.availablePermits() == 0) {
+            return; // fast path to avoid additional tasks: a worker is already draining the queue
+        }
+        // create a new worker to drain the current queue
+        Fawe.instance().getQueueHandler().submit(() -> drainQueue(false));
```

Comment:
We should probably have a synchronized block here, with a check before and after synchronizing to definitely prevent multiple submissions of drainQueue